### ISSUE TITLE
Macro expands to a crap???

### ISSUE THREAD

**Issue by funny-falcon** [crystal-lang/crystal#6254]
```
$ crystal version
Crystal 0.25.0 [7fb783f7a] (2018-06-11)

LLVM: 4.0.0
Default target: x86_64-unknown-linux-gnu
```
```crystal
struct H1
  property size : Int32 = 0
  def runtime_loop(i)
    HSIZES.each do |sz|
      if sz < @size
        ix = i / sz
        p [ix, i, sz, i / sz]
        i &= sz - 1
      end
    end
    i
  end

  def compiletime_loop(i)
    {% for sz in HSIZES %}
      if {{sz}} < @size
        ix = i / {{sz}}
        p [ix, i, {{sz}}, i / {{sz}}]
        i &= {{sz}} - 1
      end
    {% end %}
    i
  end

  HSIZES = [1<<7, 8]
end

h = H1.new
h.size = 129
puts "runtime loop"
p h.runtime_loop(1)
puts "compile time loop"
p h.compiletime_loop(1)
```
```
$ crystal run tmp/h1.cr 
runtime loop
[0, 1, 128, 0]
[0, 1, 8, 0]
1
compile time loop
[128, 1, 128, 128]
[0, 0, 8, 0]
0
```

**Comment by bew** [comment]
Adding parentheses fixes it: https://carc.in/#/r/4djp
```cr
struct H1
  def runtime_loop(i)
    HSIZES.each do |sz|
      ix = i / sz
      p [ix, i, sz, i / sz]
      i &= sz - 1
    end
    i
  end

  def compiletime_loop(i)
    {% for sz in HSIZES %}
      ix = i / ({{sz}})
      p [ix, i, ({{sz}}), i / ({{sz}})]
      i &= ({{sz}}) - 1
    {% end %}
    i
  end

  HSIZES = [1<<7, 8]
end

h = H1.new
puts "runtime loop"
p h.runtime_loop(1)
puts "compile time loop"
p h.compiletime_loop(1)
```
In the macro loop, `1<<7` is expanded literally as `1<<7`, and it can mess up the operation priorities I guess.

**Comment by funny-falcon** [comment]
@rx14 points that probably `1<<7` expanded as is , so `1/{{sz}}` expanded to `1/1<<7` . Bracketing `({{sz}})` fixes my use case.

**Comment by vlazar** [comment]
Adding {{debug}} before macro {% end %} reveals:
```
      if 1 << 7 < @size
        ix = i / 1 << 7
        p [ix, i, 1 << 7, i / 1 << 7]
        i &= 1 << 7 - 1
      end


      if 1 << 7 < @size
        ix = i / 1 << 7
        p [ix, i, 1 << 7, i / 1 << 7]
        i &= 1 << 7 - 1
      end


      if 8 < @size
        ix = i / 8
        p [ix, i, 8, i / 8]
        i &= 8 - 1
      end
```

**Comment by RX14** [comment]
We should probably automatically bracket this, would pasting in any expression with an operator at the root be a good heuristic?

**Comment by Fryguy** [comment]
> We should probably automatically bracket this

Are there cases where the caller may not want it bracketed?  It's easier to add parens than remove them.

**Comment by funny-falcon** [comment]
probably it should be bracketed not on paste time, but when expression is taken from constant. Programmer counts constant as if it has value, not an expression. So bracketing constant expression looks to be intuitive.

**Comment by straight-shoota** [comment]
So the core issue here is this:
```cr
macro foo(value)
  {{value}} / 2 # expands to 3 + 1 / 2
end

foo 3 + 1 # => 3.5
```

You would expect `{{ value }}` to expand to a single, self-contained expression. But it doesn't. It just inserts the macro id into the source code

But in other cases, this is actually expected behaviour because you expect the interpolated expression to be embedded in the expanded context.

```
macro foo(each)
  {{ each }} do |item|
    break item
  end # expands to: [1, 2, 3].each do |item| break item end
end

foo [1, 2, 3].each # => 1
```

I don't see how this could be solved generically without restricting either use case.
So in my opinion this needs to be explicitly handled at either macro definition or at the call site. In the OP's example it is definitely the macro def, the expanded macro expression should be placed in parenthesis to avoid misinterpretation.

**Comment by bew** [comment]
An option would be to add a method similar to `id` to ensure the expansion generates a self contained expression, adding parentheses if necessary.

**Comment by straight-shoota** [comment]
@bew I don't think this would add any benefit over just wrapping the expansion in parenthesis. No need for a new concept when there's already an existing one to deal with this.

**Comment by RX14** [comment]
FWIW I wouldn't mind if example 2 didn't work, or didn't without making it explicit. Example 2 is tricky to tell whats going on! `.each` and `.each(&)` are different methods!

**Comment by straight-shoota** [comment]
One of the strong points of Crystal's macro system is its simplicity. It's just plain string interpolation. The individual components it combines together don't need to be valid Crystal code themselves, only the result counts. That allows powerful features like stitching together `when` branches or constructing identifiers from macro expansions.

https://github.com/crystal-lang/crystal/blob/33670931a7ac56fa3e7b737957a9c6413605eb83/src/primitives.cr#L103

That being said, it might be a common understanding that macro expansions are self contained. For many use cases it's probably the expected behaviour. So it might not be a bad idea to avoid surprises by wrapping the expanded code in parenthesis. And that's really only required if it's not a self-contained value itself. So that should only affect some values, for example multiple expressions, calls to operators or a range literal.

**Comment by RX14** [comment]
I don't think it's just plain string interpolation though, since the incoming nodes are parsed and the `to_s` of AST nodes is often different to how it was written in the first place. They also need to be valid crystal syntax. Just not compile.

But I think focussing on what people expect is best, and I think expressions being self-contained is what people would expect?

**Comment by asterite** [comment]
Maybe we can be smart? If it's a math operation then we use parentheses. Then we'd have to see it case by case, but it might be the better option.

**Comment by asterite** [comment]
Actually, I think the current state is fine. What if you want to get the string equivalent of an expression? It would be pretty annoying to always have it inside parentheses. Right now you can control that very easily.

**Comment by straight-shoota** [comment]
Being smart might actually work. But it might be a bit suprising to see some expressions get parenthesized and others does not. So not 100% sure. And leaving things as they are is certainly okay.

It could be worth to try it at some time.

**Comment by HertzDevil** [comment]
Inside a macro or a `{% begin %}` it is indeed plain string interpolation. But bare macro interpolations must be self-contained expressions, and somehow they stay so after interpolation:

```crystal
puts({{ "3 + 1".id }} / 2) # => 2.0

macro foo
  puts({{ "3 + 1".id }} / 2)
end

foo # => 3.5
```

That said, in the first case it would be impossible to produce a `Call` node within the macro expression.
